src/render.jl

```diff
@@ -5,72 +5,169 @@ parsedocs(ds::Docile.Interface.Docs{:md}) = Markdown.parse(data(ds))
+end
+
+function push!(ents::Entries, modulename::Module, obj, ent::AbstractEntry,
+                                anchorname::AbstractString,  cat::Symbol)
+    if isexported(modulename, obj)
+        push!(ents.exported[cat], (modulename, obj, ent, anchorname))
+    elseif ents.include_internal
+        push!(ents.internal[cat], (modulename, obj, ent, anchorname))
+    end
+end
+
+type Index
+    entries::Vector{Entries}
+end
+Index() = Index(Vector{Entries}[])
+
+function update!(index::Index, ents::Entries)
+    push!(index.entries, ents)
+end
+
+function mainsetup(io::IO, mime::MIME, doc::Metadata, ents::Entries,
+                              filepath::AbstractString, config::Config)
+    # Root may be a file or directory. Get the dir.
+    rootdir = isfile(root(doc)) ? dirname(root(doc)) : root(doc)
+    for file in manual(doc)
+        writemime(io, mime, readall(joinpath(rootdir, file)))
+    end
+    # TODO: this index should be renamed to not confuse with the `Index` type
+    index = Dict{Symbol, Any}()
+    for (obj, entry) in entries(doc)
+        addentry!(index, obj, entry)
+    end
+
+    if !isempty(index)
+        ents = prepare_entries(index, doc, config)
+        if (has_items(ents.exported) || has_items(ents.internal))
+            push!(ents.sourcepaths, abspath(filepath))
+            push!(ents.modulenames, string(modulename(doc)))
+            process_entries(io, mime, "Exported", ents.exported, config)
+            process_entries(io, mime, "Internal", ents.internal, config)
+        end
+    end
+    return ents
+end
+
+# TODO: this index should be renamed to not confuse with the `Index` type
+function prepare_entries(index::Dict{Symbol, Any}, doc::Metadata, config::Config)
+    ents = Entries(config)
+    pageanchors = Dict{Symbol, Dict{String, Int}}([])
+    for k in config.category_order
+        haskey(index, k) || continue
+        k in pageanchors || (pageanchors[k] = Dict([]))
+        basenames = pageanchors[k]
+        for (s, obj) in index[k]
+            ent = entries(doc)[obj]
+            if k == :comment
+                basename = "comment"
+            else
+                # adjust basename to be a valid html id
+                basename = generate_html_id(string(k == :macro ? macroname(ent) : name(obj)))
+            end
+            basename in keys(basenames)                  ?
+                    anchornum = basenames[basename] += 1 :
+                    anchornum = basenames[basename] = 1
+            # string(k, "_", generate_html_id(s))
+            string(k, "_", generate_html_id(s))
+            push!(ents, modulename(doc), obj, entries(doc)[obj],
+                    "$(string(k))__$(basename).$(anchornum)", k)
+        end
+    end
+    return ents
 end
 
 file"docs/save.md"
-function save(file::AbstractString, modulename::Module; args...)
-    config = Config(; args...)
+function save(file::AbstractString, modulename::Module, config::Config; args...)
```

Comment:
Having an optional argument for `config`, ie. `config::Config = Config()`, here and in the other `save` method below would avoid needing to *always* provide a `Config` object. That would result in slightly cleaner user build scripts if they don't want to change any settings.